raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -2317,7 +2318,7 @@ public void resign(int epoch) {
                 "current epoch {}", epoch, currentEpoch);
             return;
         } else if (!leaderAndEpoch.isLeader(quorum.localIdOrThrow())) {
-            throw new IllegalArgumentException("Cannot resign from epoch " + epoch +
+            throw new NotLeaderException("Cannot resign from epoch " + epoch +
```

Comment:
In a similar vein, I think we can keep this as `IllegalArgumentException`. The client has attempted to resign from an epoch in which this node is not the leader.